### PR TITLE
[Graphbolt] Graphbolt Node Classification Example

### PR DESCRIPTION
**Pull request by Skeleton003**
## Description

A draft for Graphbolt node-classification example.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [x] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: examples/sampling/graphbolt/node_classification.py [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+
+def main(args):
+    if not torch.cuda.is_available():
+        args.device = "cpu"
+    print(f"Training in {args.device} mode.")
+
+    # Load and preprocess dataset.
+    dataset = gb.OnDiskDataset(
+        "/home/ubuntu/example_ogbn_products"
+    ).load()  # Please edit it according to the path of your own dataset.
```
Comment (frozenbugs) [inline examples/sampling/graphbolt/node_classification.py:298]: Let's switch to BuiltinDataset

File: examples/sampling/graphbolt/node_classification.py [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+    train_set = dataset.tasks[0].train_set
+    valid_set = dataset.tasks[0].validation_set
+    args.fanout = list(map(int, args.fanout.split(",")))
+
+    num_classes = dataset.tasks[0].metadata["num_classes"]
+
+    in_size = features.read("node", None, "feat").shape[
+        1
+    ]  # Size of feature of a single node.
+    hidden_size = 128  # Customized by you.
```
Comment (frozenbugs) [inline examples/sampling/graphbolt/node_classification.py:311]: Avoid using personal pronouns (I, you, he) in the code.
```suggestion
    hidden_size = 128
```

File: examples/sampling/graphbolt/node_classification.py [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+    out_size = num_classes  # Number of classes.
```
Comment (frozenbugs) [inline examples/sampling/graphbolt/node_classification.py:312]: Redundant comment, no extra info than the code. 
```suggestion
    out_size = num_classes
```

File: examples/sampling/graphbolt/node_classification.py [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+    model = SAGE(in_size, hidden_size, out_size)
+
+    # Model training.
+    print("Training...")
+    train(args, graph, features, train_set, valid_set, num_classes, model)
+
+    # Test the model.
+    print("Testing...")
+    test_set = dataset.tasks[0].test_set
+    test_mrr = evaluate(
```
Comment (keli-wen) [inline examples/sampling/graphbolt/node_classification.py:323]: The metric of node classification is `accuracy` rather than `mrr`. Please check your code.

File: examples/sampling/graphbolt/node_classification.py [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+        args, model, graph, features, itemset=test_set, num_classes=num_classes
+    )
+    print(f"Test MRR {test_mrr.item():.4f}")
```
Comment (keli-wen) [inline examples/sampling/graphbolt/node_classification.py:326]: Same as above.

File: examples/sampling/graphbolt/node_classification.py [hunk 6]
```diff
@@ -0,0 +1,331 @@
...
+
+
+def train(args, graph, features, train_set, valid_set, num_classes, model):
+    optimizer = torch.optim.Adam(model.parameters(), lr=args.lr)
+    dataloader = create_dataloader(
+        args, graph, features, train_set, is_train=True
+    )
+
+    for epoch in tqdm.trange(args.epochs):
+        # print(f"The beginning of epoch {epoch}")
```
Comment (keli-wen) [inline examples/sampling/graphbolt/node_classification.py:210]: Delete the useless comment.

File: examples/sampling/graphbolt/node_classification.py [hunk 7]
```diff
@@ -0,0 +1,331 @@
...
+
+    ############################################################################
+    # [Step-2]:
+    # self.sample_uniform_negative()
+    # [Note]:
+    # some extra operations, such as initializing the UniformNegativeSampler for
+    # negative sampling, may be needed in other tasks when it is not training,
+    # but not in node classification. Therefore we omit this step here.
+    ############################################################################
+    if is_train:
```
Comment (keli-wen) [inline examples/sampling/graphbolt/node_classification.py:115]: For node classification, this comment can be removed, as `sample_uniform_negative` is outside the scope of this example. Please make the corresponding update to "Step-X".

File: examples/sampling/graphbolt/node_classification.py [hunk 8]
```diff
@@ -0,0 +1,331 @@
...
+        for layer_idx, (layer, block) in enumerate(zip(self.layers, blocks)):
+            hidden_x = layer(block, hidden_x)
+            is_last_layer = layer_idx == len(self.layers) - 1
+            if not is_last_layer:
+                hidden_x = F.relu(hidden_x)
+                hidden_x = self.dropout(hidden_x)
+        return hidden_x
+
+
+"""HIGHLIGHT"""
```
Comment (frozenbugs) [inline examples/sampling/graphbolt/node_classification.py:74]: This is a little bit confusing.

File: examples/sampling/graphbolt/node_classification.py (hunk 6 above)
Comment (frozenbugs) [inline examples/sampling/graphbolt/node_classification.py:210]: remove all debug prints

File: examples/sampling/graphbolt/node_classification.py [hunk 9]
```diff
@@ -0,0 +1,331 @@
...
+    )
+
+    for epoch in tqdm.trange(args.epochs):
+        # print(f"The beginning of epoch {epoch}")
+        model.train()
+        total_loss = 0
+        for step, data in tqdm.tqdm(enumerate(dataloader)):
+            # print(f"epoch: {epoch}, step: {step}")
+
+            # blocks = data.to_dgl_blocks()
```
Comment (frozenbugs) [inline examples/sampling/graphbolt/node_classification.py:216]: remove the code if you don't need it.

File: examples/sampling/graphbolt/node_classification.py [hunk 10]
```diff
@@ -0,0 +1,331 @@
...
+
+            # The input features from the source nodes in the first layer's
+            # computation graph.
+            # x = blocks[0].srcdata["feat"]
```
Comment (frozenbugs) [inline examples/sampling/graphbolt/node_classification.py:220]: ditto.

File: examples/sampling/graphbolt/node_classification.py [hunk 11]
```diff
@@ -0,0 +1,331 @@
...
+            x = data.node_features["feat"].float()
```
Comment (frozenbugs) [inline examples/sampling/graphbolt/node_classification.py:221]: Instead of converting the data to float, better convert the model the designated data type.

File: examples/sampling/graphbolt/node_classification.py [hunk 12]
```diff
@@ -0,0 +1,331 @@
...
+
+    ############################################################################
+    # [Step-5]:
+    # self.copy_to()
+    # [Input]:
+    # 'device': The device to copy the data to.
+    # [Output]:
+    # A CopyTo object to copy the data to the specified device.
+    ############################################################################
+    datapipe = datapipe.copy_to(device=args.device)
```
Comment (frozenbugs) [inline examples/sampling/graphbolt/node_classification.py:156]: If we are not using copy_to, remove it from this PR.
